### PR TITLE
Add validation for git metadata env vars

### PR DESCRIPTION
**Pull request by ericlaz**
### What does this PR do?
Provide validation for env vars like `DD_GIT_REPOSITORY_URL` and `DD_GIT_COMMIT_SHA` in order to simplify the user-provided Git metadata input process.

### Motivation
Reduce confusion and inform users if user-provided metadata is invalid.
### Plugin Checklist

- [x] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/documentation/blob/master/content/en/tracing/trace_collection/library_config/nodejs.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

File: packages/dd-trace/src/plugins/util/user-provided-git.js [hunk 1]
```diff
@@ -39,6 +41,39 @@ function filterSensitiveInfoFromRepository (repositoryUrl) {
...
+  const size = Object.keys(metadata).length
+  if (size > 0) {
+    if (!validateGitRepositoryUrl(metadata[GIT_REPOSITORY_URL])) {
+      log.error('DD_GIT_COMMIT_SHA must be a full-length git SHA')
+    }
+    if (!validateGitCommitSha(metadata[GIT_COMMIT_SHA])) {
+      log.error('DD_GIT_REPOSITORY_URL must be a valid URL')
+    }
+  }
+}
```
Comment (juan-fernandez) [inline packages/dd-trace/src/plugins/util/user-provided-git.js:75]: logging an error is a good idea, but we want to end up doing is removing the invalid values. We could do something like what we do with `removeEmptyValues`
```suggestion
function removeInvalidGitMetadata (metadata) {
  return Object.keys(metadata).reduce((filteredTags, tag) => {
    if (tag === GIT_REPOSITORY_URL) {
      if (!validateGitRepositoryUrl(metadata[GIT_REPOSITORY_URL])) {
        log.error('DD_GIT_COMMIT_SHA must be a full-length git SHA')
        return filteredTags
      }
    }
    if (tag === GIT_COMMIT_SHA) {
      if (!validateGitCommitSha(metadata[GIT_COMMIT_SHA])) {
        log.error('DD_GIT_REPOSITORY_URL must be a valid URL')
        return filteredTags
      }
    }
    filteredTags[tag] = metadata[tag]
    return filteredTags
  }, {})
}
```

File: packages/dd-trace/src/plugins/util/user-provided-git.js [hunk 2]
```diff
@@ -75,6 +110,8 @@ function getUserProviderGitMetadata () {
     [GIT_COMMIT_AUTHOR_EMAIL]: DD_GIT_COMMIT_AUTHOR_EMAIL,
     [GIT_COMMIT_AUTHOR_DATE]: DD_GIT_COMMIT_AUTHOR_DATE
   })
+  validateGitMetadata(metadata)
+  return metadata
```
Comment (juan-fernandez) [inline packages/dd-trace/src/plugins/util/user-provided-git.js:114]: with the above suggestion, we could do 
```suggestion
  return removeInvalidGitMetadata(metadata)
```

File: packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js [hunk 3]
```diff
@@ -202,7 +202,57 @@ describe('git_metadata', () => {
       done()
     })
   })
+  it('should return false if Git repository URL is invalid', (done) => {
```
Comment (juan-fernandez) [inline packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js:205]: `done` is only used when tests are asynchronous, but this is not the case here, so we can remove it 
```suggestion
  it('should return false if Git repository URL is invalid', () => {
```

File: packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js [hunk 4]
```diff
@@ -202,7 +202,57 @@ describe('git_metadata', () => {
...
+    const invalidUrl3 = 'http://test.com/repo/dummy.4git'
+    const invalidUrl4 = 'https://test.com/repo/dummy.gi'
+    const invalidUrl5 = 'www.test.com/repo/dummy.git'
+    const invalidUrl6 = 'test.com/repo/dummy.git'
+
+    const invalidUrls = [invalidUrl1, invalidUrl2, invalidUrl3, invalidUrl4, invalidUrl5, invalidUrl6]
+    invalidUrls.forEach((invalidUrl) => {
+      expect(validateGitRepositoryUrl(invalidUrl)).to.be.false
+    })
+    done()
```
Comment (juan-fernandez) [inline packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js:217]:
```suggestion
```

File: packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js [hunk 5]
```diff
@@ -202,7 +202,57 @@ describe('git_metadata', () => {
...
+  })
+  it('should return true if Git repository URL is valid', (done) => {
```
Comment (juan-fernandez) [inline packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js:219]:
```suggestion
  it('should return true if Git repository URL is valid', () => {
```

File: packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js [hunk 6]
```diff
@@ -202,7 +202,57 @@ describe('git_metadata', () => {
...
+    const validUrl2 = 'http://test.com/repo/dummy.git'
+    const validUrl3 = 'https://github.com/DataDog/dd-trace-js.git'
+    const validUrl4 = 'git@github.com:DataDog/dd-trace-js.git'
+    const validUrl5 = 'git@github.com:user/repo.git'
+
+    const validUrls = [validUrl1, validUrl2, validUrl3, validUrl4, validUrl5]
+    validUrls.forEach((validUrl) => {
+      expect(validateGitRepositoryUrl(validUrl)).to.be.true
+    })
+    done()
```
Comment (juan-fernandez) [inline packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js:230]:
```suggestion
```

File: packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js [hunk 7]
```diff
@@ -202,7 +202,57 @@ describe('git_metadata', () => {
...
+    const invalidSha3 = ''
+    const invalidSha4 = 'test'
+    const invalidSha5 = 'cb466452bfe18d4f6be2836c2a5551843013cf382cb466452bfe18d4f6be2836c2a5551843013cf382'
+    const invalidSha6 = 'cb466452bfe18d4f6be2836c2a5551843013cf2'
+    const invalidSha7 = 'cb466452bfe18d4f6be2836c2a5551843013cf3812342239203182304928'
+    const invalidShas = [invalidSha1, invalidSha2, invalidSha3, invalidSha4, invalidSha5, invalidSha6, invalidSha7]
+    invalidShas.forEach((invalidSha) => {
+      expect(validateGitCommitSha(invalidSha)).to.be.false
+    })
+    done()
```
Comment (juan-fernandez) [inline packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js:244]:
```suggestion
```

File: packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js [hunk 8]
```diff
@@ -202,7 +202,57 @@ describe('git_metadata', () => {
...
+    const validUrl4 = 'git@github.com:DataDog/dd-trace-js.git'
+    const validUrl5 = 'git@github.com:user/repo.git'
+
+    const validUrls = [validUrl1, validUrl2, validUrl3, validUrl4, validUrl5]
+    validUrls.forEach((validUrl) => {
+      expect(validateGitRepositoryUrl(validUrl)).to.be.true
+    })
+    done()
+  })
+  it('should return false if Git commit SHA is invalid', (done) => {
```
Comment (juan-fernandez) [inline packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js:232]:
```suggestion
  it('should return false if Git commit SHA is invalid', () => {
```

File: packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js [hunk 9]
```diff
@@ -202,7 +202,57 @@ describe('git_metadata', () => {
...
+    const invalidSha5 = 'cb466452bfe18d4f6be2836c2a5551843013cf382cb466452bfe18d4f6be2836c2a5551843013cf382'
+    const invalidSha6 = 'cb466452bfe18d4f6be2836c2a5551843013cf2'
+    const invalidSha7 = 'cb466452bfe18d4f6be2836c2a5551843013cf3812342239203182304928'
+    const invalidShas = [invalidSha1, invalidSha2, invalidSha3, invalidSha4, invalidSha5, invalidSha6, invalidSha7]
+    invalidShas.forEach((invalidSha) => {
+      expect(validateGitCommitSha(invalidSha)).to.be.false
+    })
+    done()
+  })
+  it('should return true if Git commit SHA is valid', (done) => {
```
Comment (juan-fernandez) [inline packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js:246]:
```suggestion
  it('should return true if Git commit SHA is valid', () => {
```

File: packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js [hunk 10]
```diff
@@ -202,7 +202,57 @@ describe('git_metadata', () => {
...
+    const validSha1 = 'cb466452bfe18d4f6be2836c2a5551843013cf38'
+    const validSha2 = 'cb466452bfe18d4f6be2836c2a5551843013cf381234223920318230492823f3'
 
+    const validShas = [validSha1, validSha2]
+    validShas.forEach((validSha) => {
+      expect(validateGitCommitSha(validSha)).to.be.true
+    })
+    done()
```
Comment (juan-fernandez) [inline packages/dd-trace/test/ci-visibility/exporters/git/git_metadata.spec.js:254]:
```suggestion
```

**Review by juan-fernandez — COMMENTED** [review]
great job! And tests look very good 😄

File: packages/dd-trace/src/plugins/util/user-provided-git.js [hunk 11]
```diff
@@ -39,6 +41,35 @@ function filterSensitiveInfoFromRepository (repositoryUrl) {
   }
 }
 
+function validateGitRepositoryUrl (repoUrl) {
+  return /((http|git|ssh|http(s)|file|\/?)|(git@[\w.]+))(:(\/\/)?)([\w.@:/\-~]+)(\.git)(\/)?/.test(repoUrl)
```
Comment (uurien) [inline packages/dd-trace/src/plugins/util/user-provided-git.js:45]: why is a group the `s` of `http(s)`?
and this regex accepts things like `/://aaa.git` is this expected?

File: packages/dd-trace/src/plugins/util/user-provided-git.js [hunk 12]
```diff
@@ -39,6 +41,35 @@ function filterSensitiveInfoFromRepository (repositoryUrl) {
...
+  return Object.keys(metadata).reduce((filteredTags, tag) => {
+    if (tag === GIT_REPOSITORY_URL) {
+      if (!validateGitRepositoryUrl(metadata[GIT_REPOSITORY_URL])) {
+        log.error('DD_GIT_COMMIT_SHA must be a full-length git SHA')
+        return filteredTags
+      }
+    }
+    if (tag === GIT_COMMIT_SHA) {
+      if (!validateGitCommitSha(metadata[GIT_COMMIT_SHA])) {
+        log.error('DD_GIT_REPOSITORY_URL must be a valid URL')
```
Comment (uurien) [inline packages/dd-trace/src/plugins/util/user-provided-git.js:64]: Mixed log messages:
```suggestion
        log.error('DD_GIT_REPOSITORY_URL must be a valid URL')
        return filteredTags
      }
    }
    if (tag === GIT_COMMIT_SHA) {
      if (!validateGitCommitSha(metadata[GIT_COMMIT_SHA])) {
        log.error('DD_GIT_COMMIT_SHA must be a full-length git SHA')
```

**Review by uurien — CHANGES_REQUESTED** [review]

File: LICENSE-3rdparty.csv [hunk 13]
```diff
@@ -67,5 +67,6 @@ dev,sinon,BSD-3-Clause,Copyright 2010-2017 Christian Johansen
 dev,sinon-chai,WTFPL and BSD-2-Clause,Copyright 2004 Sam Hocevar 2012–2017 Domenic Denicola
 dev,tap,ISC,Copyright 2011-2022 Isaac Z. Schlueter and Contributors
 dev,tape,MIT,Copyright James Halliday
-file,aws-lambda-nodejs-runtime-interface-client,Apache 2.0,Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (juan-fernandez) [inline LICENSE-3rdparty.csv:70]: unrelated fix. This allows github to render this nicely: https://github.com/DataDog/dd-trace-js/blob/eric-navarro/validate-git-metadata/LICENSE-3rdparty.csv

**Review by uurien — APPROVED** [review]

**Review by juan-fernandez — APPROVED** [review]

**Pull request merged**